### ISSUE TITLE
Exception Code to be part of DatabaseException

### ISSUE THREAD

**Issue by dhavalkshah** [tekartik/sqflite#56]
Hi,

I wanted to write an UPSERT using sqflite 
My approach was to try to insert and if insert fails with "UNIQUE constraint failed" I will try to update
To catch the error i wrote something like:

```
try{
    await db.rawInsert('INSERT INTO MY_TABLE (col1 , col2) VALUES ( ?, ?)',['value1','value2']);
}
on DatabaseException catch (e){
    print(e);
}
```
Now, the insert fails as expected and i get the print as 
```
DatabaseException(UNIQUE constraint failed:....
(Sqlite code 1555), (OS error - 2:No such file or directory)) sql.....
......
```
The exception object does not have a code as a member. 
Comparing that with (in this case with 1555) would have helped me to take appropriate action, fire an update. 
I would not like to compare message strings. 

It would have been great if the sqlite error code could have been part of the derived class viz. **DatabaseException**

Regards
Dhaval

**Comment by alextekartik** [comment]
Hi and thanks for your report. I understand the concern. Unfortunaly getting the sqlite result code is not consistent on iOS and Android. For your case, iOS will report code 19 while android code 1555 (for primary key) or 2067. And even worse, getting the error code is not consistent neither and most of the time requires parsing the text message from sqlite (at least on Android)

So far my case-by-case solution was to add helpers in DatabaseException from user requests. I added a "isUniqueConstraintError" in a version 0.8.8 I just published. Does it fit your need? At least I know this method it will work on both iOS and Android (while getting the result code - which exists but is not public - only works on Android so far). The more I get examples from user on the text they get the more I can refine the exception handling which is definitely hacky and a pain (even when using it natively on both platforms)

**Comment by dhavalkshah** [comment]
Hi @alextekartik ,  

Thanks for the update
Yes, isUniqueConstraintError will work

> iOS will report code 19 while android code 1555 (for primary key) or 2067. 

Oh, I never knew that the error codes are different for android or iOS. I always thought that it is the underlying sqlite error code which should have been consistent, irrespective of the underlying OS.

I will surely try the latest version and let you know 
Also, I will provide you the exact string that I encountered, if required

> The more I get examples from user on the text they get the more I can refine the exception handling which is definitely hacky and a pain (even when using it natively on both platforms)

I will surely do my bit of providing you all that I can to help the larger community

Thanks & Regards
Dhaval

**Comment by dhavalkshah** [comment]
Hi @alextekartik 

Tried isUniqueConstraintError. It works as expected
Thus closing this issue.
Appreciate the help

~Dhaval

P.S.: In my application (in C, with Oracle as DB) I have encapsulated following errors/exceptions:
DUP_REC -> Duplicate Record. Equivalent to isUniqueConstraintError
NO_REC -> No records fetched
LOCK_WAIT -> Failed SELECT FOR UPDATE NO WAIT for wait threshold time (I avoid SELECT FOR UPDATE WAIT)
LOCK_NO_WAIT -> Failed SELECT FOR UPDATE NO WAIT without threshold
SYNTAX_ERR -> Equivalent to isSyntaxError 

You might want to look at implementing some of them
